rfd/0086-stream-audit-events.md

```diff
@@ -0,0 +1,227 @@
+> If both `StartingSequenceNumber` and `EndingSequenceNumber` are present, then that shard is closed and can no longer receive more data.
+
+For each of these shards, we also retrieve a shard iterator  using [`GetShardIterator`], providing the following information:
+
+```json
+{
+   "SequenceNumber": "string",
+   "ShardId": "string",
+   "ShardIteratorType": "string",
+   "StreamArn": "string"
+}
+```
+
+We have the following `ShardIteratorType`s:
+
+> - `AT_SEQUENCE_NUMBER` - Start reading exactly from the position denoted by a specific sequence number.
+> - `AFTER_SEQUENCE_NUMBER` - Start reading right after the position denoted by a specific sequence number.
+> - `TRIM_HORIZON` - Start reading at the last (untrimmed) stream record, which is the oldest record in the shard. In DynamoDB Streams, there is a 24 hour limit on data retention. Stream records whose age exceeds this limit are subject to removal (trimming) from the stream.
+> - `LATEST` - Start reading just after the most recent stream record in the shard, so that you always read the most recent data in the shard.
+
+Given a shard id `$ID` (returned by `DescribeStream`), if `streamCursor.ShardIdToSequenceNumber` contains `$ID`, then we set the `ShardIteratorType` to `AFTER_SEQUENCE_NUMBER` and `SequenceNumber` to `streamCursor.ShardIdToSequenceNumber[$ID]`.
+Otherwise, we can either set it to `TRIM_HORIZON` or to `LATEST`.
+In this case, we will set it to `LATEST` to maintain the current behaviour in `shards.go`.
+
+Once we have a shard iterator returned by `GetShardIterator`, we can finally use it to [`GetRecords`] from the stream.
+
+[`dynamodb.Log.SearchEvents`]: https://github.com/gravitational/teleport/blob/cf205b01a5aa88fd4fcdb499ff9b9c40c4e5c335/lib/events/dynamoevents/dynamoevents.go#L558-L560
+[checkpoint key]: https://github.com/gravitational/teleport/blob/cf205b01a5aa88fd4fcdb499ff9b9c40c4e5c335/lib/events/dynamoevents/dynamoevents.go#L538-L548
+[`DescribeStream`]: https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_streams_DescribeStream.html
+[`GetShardIterator`]: https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_streams_GetShardIterator.html
+[`GetRecords`]: https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_streams_GetRecords.html
+
+The steps described above are already implemented in [`lib/backend/dynamo/shards.go`].
+The missing feature is allowing the stream to be resumed by providing a stream cursor.
+
+### Additional notes
+- Streaming of audit events requires that DynamoDB streams are enabled for the events table (just like they're [enabled for the backend]).
+As this is only needed for Teleport Cloud, the tenant-operator will be responsible for enabling this.
+- There's a [limit on the number of simultaneous readers] of stream shard. This limit is 2, which is enough for MAU as we plan to stream audit events from a single process.
```

Comment:
These  streaming audit events functionality will be used only by another in depended Cloud  process not related to Teleport proxy/auth service ?